streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -131,45 +94,85 @@ public void register(final StateRestorer restorer) {
-                    restorer.setRestoredOffset(0);
+
+        // try to fetch end offsets for the initializable restorers and remove any partitions
+        // where we already have all of the data
+        try {
+            endOffsets.putAll(consumer.endOffsets(initializable.keySet()));
+        } catch (final TimeoutException e) {
+            // if timeout exception gets thrown we just give up this time and retry in the next run loop
+            log.debug("{} Could not fetch end offset for {}; will fall back to partition by partition fetching", logPrefix, initializable);
+            initializable.clear();
+        }
+
+        if (!initializable.isEmpty()) {
+            Iterator<TopicPartition> iter = initializable.keySet().iterator();
+            while (iter.hasNext()) {
+                final TopicPartition topicPartition = iter.next();
+                final Long offset = endOffsets.get(topicPartition);
+
+                // offset should not be null; but since the consumer API does not guarantee it
+                // we add this check just in case
+                if (offset != null) {
+                    // do not need to check restorer since it will never by null
+                    final StateRestorer restorer = needsInitializing.get(topicPartition);
+                    if (restorer.checkpoint() >= offset) {
+                        restorer.setRestoredOffset(restorer.checkpoint());
+                        iter.remove();
+                    } else if (restorer.offsetLimit() == 0 || endOffsets.get(topicPartition) == 0) {
+                        restorer.setRestoredOffset(0);
+                        iter.remove();
+                    } else {
+                        restorer.setEndingOffset(offset);
+                    }
+                    needsInitializing.remove(topicPartition);
                 } else {
-                    newTasksNeedingRestoration.put(topicPartition, restorer);
-                    final Long endOffset = endOffsets.get(topicPartition);
-                    restorer.restoreStarted(restorer.startingOffset(), endOffset);
+                    iter.remove();
                 }
             }
         }
 
-        log.debug("{} Starting restoring state stores from changelog topics {}", logPrefix, newTasksNeedingRestoration.keySet());
+        // set up restorer for those initializable
+        if (!initializable.isEmpty()) {
+            startRestoration(initializable);
+        }
+
+        // if there are still some restorers whose changelog partitions are not known yet,
+        // try to fetch all metadata info once and in the next run loop we will retry again
+        if (!needsInitializing.isEmpty()) {
+            log.debug("{} Changelog partitions {} metadata are unknown so they cannot be used for restoration;" +
+                    "will retry in the next run loop", logPrefix, needsInitializing.keySet());
+            refreshChangelogInfo();
```

Comment:
Maybe we should move this to the top of the method and then remove the call of `refreshChangelogInfo()` from `TaskManager`?

Reply:
Yeah you are right. I guess I was mistakenly thinking that the `needsInitialized` can still be updated after the `initTopology`.